src/modules/ssh_docker/mappers/docker_container.ts

```diff
@@ -0,0 +1,134 @@
+    return out;
+  }
+
+  private async getDockerForServer(ctx: Context, serverName: string): Promise<Docker | undefined> {
+    const sshClient: SSH2Promise = await ctx.getSshClient(serverName);
+    const sshConfig = sshClient.config[0];
+    const modem = new Modem({
+      protocol: 'ssh',
+      host: sshConfig.host,
+      port: sshConfig.port,
+      username: sshConfig.username,
+      sshOptions: {
+        privateKey: sshConfig.privateKey,
+        passphrase: sshConfig.passphrase,
+      },
+    });
+    const docker = new Docker({ modem } as DockerOptions);
+    try {
+      await docker.ping();
+    } catch (e) {
+      return undefined;
+    }
+    return docker;
+  }
+
+  cloud = new Crud({
+    create: async (es: DockerContainer[], ctx: Context) => {
+      // create docker containers through RPC
+      await this.db.delete(es, ctx);
+    },
+    read: async (ctx: Context, id?: string) => {
+      if (!!id) {
+        const { serverName, containerId } = this.idFields(id);
+        const docker = await this.getDockerForServer(ctx, serverName);
+        if (!docker) return;
+
+        const container = (await docker.listContainers()).find(c => c.Id === containerId);
+        if (!container) return undefined;
+        return this.containerMapper(serverName, container);
+      } else {
+        const out: DockerContainer[] = [];
+        const sshCredentials: SshCredentials[] = await sshAccounts.sshCredentials.cloud.read(ctx);
+        const serverNames = sshCredentials.map(c => c.name);
+        for (const serverName of serverNames) {
+          const docker = await this.getDockerForServer(ctx, serverName);
+          if (!docker) continue;
+
+          const containers = await docker.listContainers({ all: true });
+          for (const rawContainer of containers) {
+            const container = this.containerMapper(serverName, rawContainer);
+            out.push(container);
+          }
+        }
+        return out;
+      }
+    },
+    update: async (es: DockerContainer[], ctx: Context) => {
+      const out = [];
+      for (const e of es) {
+        // restore because you can't change a container
+        const cloudRecord: DockerContainer = ctx?.memo?.cloud?.DockerContainer?.[this.entityId(e)];
+        cloudRecord.id = e.id;
+        await this.db.update(cloudRecord, ctx);
+        out.push(cloudRecord);
+      }
+      return out;
+    },
+    delete: async (es: DockerContainer[], ctx: Context) => {
+      for (const e of es) {
+        const docker = await this.getDockerForServer(ctx, e.serverName);
+        const container = await docker!.getContainer(e.containerId);
+        try {
+          await container.stop();
+        } catch (e: any) {
+          // statusCode 304 means the container is already stopped
+          if (e.statusCode !== 304) throw e;
+        }
+        // TODO: maybe we should not remove the container - just stopping would be enough?
+        await container.remove();
```

Comment:
No, I think this belongs here. There is the `status` column that would show the current status of the container. But that would require a change to the `update` method above.

Reply:
Makes sense. Implemented as you suggested. But I had to select fields such that they are in both `createContainer` and `inspectContainer` - or interchangably usable.